scalding-core/src/main/scala/com/twitter/scalding/typed/WritePartitioner.scala

```diff
@@ -215,80 +215,163 @@ object WritePartitioner {
       }
     }
 
+    /**
+     * If cascading would conside the current pipe as a Logical reduce
+     * we can avoid some forces below
+     */
+    @annotation.tailrec
+    def isLogicalReduce(tp: TypedPipe[Any]): Boolean = {
+      import TypedPipe._
+      tp match {
+        case EmptyTypedPipe | IterablePipe(_) | SourcePipe(_) => false
+        case CounterPipe(a) => isLogicalReduce(a)
+        case cp@CrossPipe(_, _) => isLogicalReduce(cp.viaHashJoin)
+        case cp@CrossValue(_, _) => isLogicalReduce(cp.viaHashJoin)
+        case DebugPipe(p) => isLogicalReduce(p)
+        case FilterKeys(p, _) => isLogicalReduce(p)
+        case Filter(p, _) => isLogicalReduce(p)
+        case FlatMapValues(p, _) => isLogicalReduce(p)
+        case FlatMapped(p, _) => isLogicalReduce(p)
+        case ForceToDisk(_) => false // not reducers for sure, could be a map-only job
+        case Fork(_) => false // TODO, not super clear
+        case HashCoGroup(left, _, _) => isLogicalReduce(left)
+        case MapValues(p, _) => isLogicalReduce(p)
+        case Mapped(p, _) => isLogicalReduce(p)
+        case MergedTypedPipe(_, _) => false
+        case ReduceStepPipe(_) => true
+        case SumByLocalKeys(p, _) => isLogicalReduce(p)
+        case TrappedPipe(p, _, _) => isLogicalReduce(p)
+        case CoGroupedPipe(_) => true
+        case WithOnComplete(p, _) => isLogicalReduce(p)
+        case WithDescriptionTypedPipe(p, _) => isLogicalReduce(p)
+      }
+    }
+
+    /**
+     * This is a lattice value that tracks
+     * what we have seen below a given TypedPipe as
+     * we recurse up.
```

Comment:
This is super neet/nice, could you expand the comments here -- don't think i'd have followed other than I know what your aiming to do.

(Motivation/why we are tracking/what we are aiming to do with it)